src/main/kotlin/cacophony/grammars/FindFirst.kt

```diff
@@ -1,10 +1,62 @@
 package cacophony.grammars
 
 import cacophony.automata.DFA
+import cacophony.utils.getProperTransitiveClosure
+
+data class GeneralizedSymbol<StateType, SymbolType, ResultType>(
+    val state: DFAStateReference<StateType, SymbolType, ResultType>?,
+    val symbol: SymbolType?,
+) {
+    init {
+        if (state == null && symbol == null) {
+            throw IllegalArgumentException("State and Symbol cannot both be nulls")
+        }
+    }
```

Comment:
you can set `state` and `symbol` to be non-nullable in class header and get rid of it

i see that if you do it, you get compilation error in two places below, but there you can do `!!` assertion and this way you achieve exactly the same as currently (if the reference actually turns out to be `null` in runtime, `KotlinNullPointerException` will be thrown)

Reply:
I don't want them to be nulls at the same time. Either one of them can be null, as long as the other one is not.